mgr/dashboard: orchestrator integration initial works

bk201 opened the pull request:
- Display hosts, inventory, and services from orchestrator
- Allow adding/removing hosts

Fixes: https://tracker.ceph.com/issues/40337
Fixes: https://tracker.ceph.com/issues/40336
Fixes: https://tracker.ceph.com/issues/38233

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -1,12 +1,79 @@
+    def delete(self, hostname):
+        orch_client = OrchClient()
+        self._check_orchestrator_host_op(orch_client, hostname, False)
+        orch_client.hosts.remove(hostname)
```
I'd say this depends on #29063 to be merged

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -1,12 +1,79 @@
+                code='orchestrator_remove_nonexistent_host',
+                msg='Remove a non-existent host {} from orchestrator'.format(
+                    hostname),
+                component='orchestrator')
```
Hm, I'd think the orchestrator cli would also benefit from these checks. Might it make sense to move them into the orch?

bk201 replied:
This is checked both in frontend and backend. Yes, it will be good if orchestrator can check this too. I should be done with another PR.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/orchestrator.py:
```diff
@@ -0,0 +1,49 @@
+    def status(self):
+        orch = OrchClient()
+        return {
+            'available': orch.available()
```
The orchestrator also provides a status message. might make sense to add it here.

bk201 replied:
fixed

sebastian-philipp commented on src/pybind/mgr/dashboard/services/exception.py:
```diff
@@ -120,3 +120,12 @@ def handle_send_command_error(component):
+def handle_orchestrator_error(component):
+    try:
+        yield
+    except RuntimeError as e:
```
See http://docs.ceph.com/docs/master/mgr/orchestrator_modules/#error-handling

Everything that is not an `OrchestratorError` should be an internal server error. 

Every exception that is raised that should be handled gracefully should be converted to an OrchestratorError in the orchestrator modules.

bk201 replied:
Looks like an exception on remote is always converted to RuntimeError.
https://github.com/ceph/ceph/blob/7265b55d094a639be50a567d3be92fba94c04786/src/pybind/mgr/mgr_module.py#L1269

sebastian-philipp replied:
Yeah, I wrote that one. Within _Completions, exceptions are carefully transferred into the current subinterpreter and then re-raised.

bk201 replied:
Thanks, I saw that. But it's weird I always catch a `RuntimeError` on an unimplemented method. I need to investigate further.

sebastian-philipp commented on src/pybind/mgr/dashboard/services/orchestrator.py:
```diff
@@ -1,38 +1,98 @@
+
+def wait_api_result(method):
+    @wraps(method)
+    def inner(*args, **kwargs):
```
```suggestion
    def inner(self, *args, **kwargs):
```
?

bk201 replied:
fixed

epuertat commented on src/pybind/mgr/dashboard/services/orchestrator.py:
```diff
@@ -1,38 +1,101 @@
+    def status(self):
         try:
-            status, desc = super(OrchClient, self).available()
+            status, desc = super(OrchestratorAPI, self).available()
```
In Python you better ask forgiveness than permission, which means: instead of checking whether the Orchestrator is available, you just run your actions, and if not available it will raise an exception.

sebastian-philipp replied:
[I agree](http://docs.ceph.com/docs/master/mgr/orchestrator_modules/#orchestrator.Orchestrator.available) but I don't see this used as such here?

epuertat replied:
Yes, you're right. I think I got mixed up here: the `if orch.available():` statements appear twice a few lines below.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/health.py:
```diff
@@ -118,7 +119,10 @@ def fs_map(self):
-        return len(mgr.list_servers())
+        all_hosts = mgr.list_servers()
+        if self._has_permissions(Permission.READ, Scope.ORCHESTRATOR):
+            all_hosts = merge_with_orchestrator(all_hosts)
```
:+1: I think this is a problem of too fine grained permissions.

epuertat replied:
Yes, I agree. I think with the current number of scopes (or even a few less ones) we are ok.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -1,12 +1,79 @@
-        return mgr.list_servers()
+        hosts = mgr.list_servers()
+        if self._has_permissions(Permission.READ, Scope.ORCHESTRATOR):
+            return merge_with_orchestrator(hosts)
```
We're not breaking existing clients. I'm :+1: with merging the hosts in here and keep the code as it is.

bk201 replied:
>    2. If we don't like the idea of issuing 2 requests, let's do the math in the back-end but at least let's make it explicit (`/host?sources=[ceph,orchestrator]`)

I prefer not adding a new endpoint and enhancing `/api/host`. How about adding a new field `sources` in each host object:

```javascript
[
  {
    "hostname": "host1",
    "sources": {
      "ceph": true,
      "orchestrator": true
    }
  },
  {
    "hostname": "host2",
    "sources": {
      "ceph": false,
      "orchestrator": true
    }
  }
]
```

By default, a GET to `/api/host` returns all hosts included orchestrator ones. And then implement a filter with `sources` parameter.

bk201 replied:
address in https://github.com/ceph/ceph/pull/29127/commits/cd101da73a4e933100ae3b3dad242c38bedcb4c1

epuertat replied:
Thanks @bk201 , I'm not too fan about extending Resources via query params (that's why I suggested first the new endpoint approach), but I think it's better than to  backward-incompatibly modify the behavior. The reasoning for me not excited about the query param option is because query params work best when they are left just to representation modifiers (pagination, filtering, field selection, sorting, etc.), and are all the same to all endpoints.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -1,12 +1,79 @@
+    return Task("host/{}".format(name), metadata, wait_for)
+
+
+def merge_with_orchestrator(mgr_hosts):
```
(agreeing with the other points)

> By doing that, I cannot easily use `merge_with_orchestrator` for just getting the orchestrator hosts (well, I can, but I need to pass an empty list, which will be copied into a separate one...).

If you want to just have the orchestrator hosts, then you can can just call the orchestrator.

bk201 replied:
> Not a big fan of functional programming, but here I wonder why not having a method of which does not modify the input list and just provides the orchestrator list of hosts?

The function does not modify input list because of `copy` code. And as @sebastian-philipp pointed out, if we need only hosts in orchestrator, just use the orchestrator client.

bk201 replied:
Sorry, I should switch to `deepcopy` for this.

sebastian-philipp commented on src/pybind/mgr/dashboard/services/orchestrator.py:
```diff
@@ -1,38 +1,101 @@
+
+        self.hosts = HostManger(self.api)
+        self.inventory = InventoryManager(self.api)
+        self.services = ServiceManager(self.api)
```
just fyi, The orchestrator interface aka `OrchestratorMixin` has no state and will never have. (well except that it needs to know the `mgr`.

sebastian-philipp replied:
If you have a better way of giving `OrchestratorClientMixin` access to the dashboard's `mgr`, feel free to change it.

bk201 replied:
> In that case I don't think it's needed to instantiate it, don't you think? Maybe it could be initialized once (singleton-like) or by a class method (`OrchClient.init()`) 

Yes. I'll refine this.

>If you have a better way of giving OrchestratorClientMixin access to the dashboard's mgr, feel free to change it.

My intention is to wrap the class that extends `OrchestratorClientMixin` in the client. Because I tend to implement methods that have similar names with methods in `OrchestratorClientMixin` (which extends `Orchestrator`). For example, `get_hosts`, `add_host`, and `remove_host`. A client that extends `OrchestratorClientMixin` make s me feel like I'm writing an orchestrator backend.

bk201 replied:
The class is initialized once now, client code use `get_client()` to retrieve the instance. Please see https://github.com/ceph/ceph/pull/29127/commits/5062a39030f513efa5e36e13c5e69a5a9bb4aa4a

epuertat replied:
Thanks!

bk201 commented on src/pybind/mgr/dashboard/controllers/health.py:
```diff
@@ -118,7 +119,10 @@ def fs_map(self):
     def host_count(self):
-        return len(mgr.list_servers())
+        all_hosts = mgr.list_servers()
+        if self._has_permissions(Permission.READ, Scope.ORCHESTRATOR):
```
I do agree with the approach that hiding the orchestrator from users.

Maybe adding a new `ORCHESTRATOR` scope is unnecessary, we can just allow adding/removing hosts via orchestrator API if a user has  `HOSTS/{CREATE|DELETE}` permission. I reviewed existing scopes again and found out they are all based on resources exported by Dashboard, so exporting a new `ORCHESTRATOR` scope might be a bit weird.

For inventory (devices), we can create a new `INVENOTRY` scope. Users with this permission can view disks' information, light disk LEDs, and operations related to physical assets.

For services, we can create a new `SERVICE` scope. Users with this permission can view services on hosts, start/stop/restart them.

For OSD operations with orchestrator involved, we use existing `OSD` scope.

If a user has permissions defined in Dashboard, he can perform orchestrator-related operations on resources (only if orchestrator is available).

rjfd replied:
> Maybe adding a new ORCHESTRATOR scope is unnecessary, we can just allow adding/removing hosts via orchestrator API if a user has HOSTS/{CREATE|DELETE} permission. I reviewed existing scopes again and found out they are all based on resources exported by Dashboard, so exporting a new ORCHESTRATOR scope might be a bit weird.

I also think the "orchestrator" should not be a scope. The orchestrator is just an implementation mechanism of the dashboard features. From the dashboard user's point of view, the orchestrator is invisible.

bk201 replied:
Fix in 9d9bcdd

epuertat replied:
@bk201 on the Inventory & Services scope, I think both things are cross-component. I'd reuse existing scopes: for example, I think Invetory/disk operations might fit well under OSD (or Host?) scope, while services should honor their corresponding component permissions (IMHO the Scope for RGW Service is more RGW than a new generic Service scope; same for CephFS, and so on).

bk201 commented:
> Why do we need a persistent shim to the orchestrator?

Switch to a class getter for single instance. (https://github.com/ceph/ceph/pull/29127/commits/1b49673d6954f70cacc0be74cacbe4aab6cbc3a4)

rjfd commented:
@bk201 I agree with @epuertat  inventory operations should use the `hosts` scope, and services should use the scope that is related to.

bk201 commented:
> @bk201 I agree with @epuertat inventory operations should use the `hosts` scope, and services should use the scope that is related to.

Thank @epuertat  and @rjfd  for comments.

Inventory is attached to hosts, so it's sane to leverage `HOSTS` scope.

As for services, it's nice to use scopes related to them too.
According to [current orchestrator implementations](https://docs.ceph.com/docs/master/mgr/orchestrator_cli/#current-implementation-status), I summarize service types in orchestrator and [existing Dashboard scopes](https://github.com/ceph/ceph/blob/1a3a1d131054b4def6113855f32f4fc48c61be07/src/pybind/mgr/dashboard/security.py#L13) might be used in the following table:

| Service in Orchestrator | Scope in Dashboard |
|---|---|
| mgr        | MANAGER       | 
| mon        | MONITOR       |
| iscsi      | ISCSI         |
| mds        | CEPHFS        |
| nfs        | NFS_GANESHA   |
| rbd-mirror | RBD_MIRRORING |
| rgw        | RGW           |

Operations for these services might be implemented in pages related to services rather than main services page. For example, if a user wants to add a `mon` service, we can add this feature in `Monitors` page (a button on monitors table). 

Another question: what to do when **listing services**? Return only those services that users have READ permission on? Or always allow read and apply permission checks when users want to perform create/delete/update operation on services?

epuertat commented:
> Another question: what to do when **listing services**? Return only those services that users have READ permission on? Or always allow read and apply permission checks when users want to perform create/delete/update operation on services?

Looks great, thanks!

On the listing services, where is that going to be displayed? If if's in come Orchestrator page, let's stick to Orch Scope; if it's in each service page (RGW, CephFS, etc), let's use each service scope. Nevertheless, it might happen that there's no easy way to handle that with the current scopes (for example, Options has a separate scope, while ideally users should only have access to the Options related to their enabled Scopes; unfortunately that's not easy to do as not all options affect single services).

bk201 commented:
> On the listing services, where is that going to be displayed?

Services of a host will be displayed on [*Service tab* of host](https://tracker.ceph.com/issues/40337?issue_count=240&issue_position=55&next_issue_id=40336&prev_issue_id=40413#Hosts-page) and a [new page](https://tracker.ceph.com/issues/40337?issue_count=240&issue_position=55&next_issue_id=40336&prev_issue_id=40413#Services-page) for all services.
Please see the links, thanks!

epuertat commented:
> > On the listing services, where is that going to be displayed?
> 
> Services of a host will be displayed on [_Service tab_ of host](https://tracker.ceph.com/issues/40337?issue_count=240&issue_position=55&next_issue_id=40336&prev_issue_id=40413#Hosts-page) and a [new page](https://tracker.ceph.com/issues/40337?issue_count=240&issue_position=55&next_issue_id=40336&prev_issue_id=40413#Services-page) for all services.
> Please see the links, thanks!

I think all users with Read perm to Hosts page (until now no other perm than Read made sense) should see all services available. However if the Hosts page is going to be used for deploying/scaling/decomissioning service nodes on a per-host basis, then we should enable them based on the corresponding services (RGW, RBD, etc).

Nonetheless, I'm thinking how to improve this, as this might end up being quite complicate to handle (with components that fully served to a specific purpose that was easy, but with complex workflows involving multiple back-end entities it's becoming more intricate).

The current approach is:
```html
<div *ngIf="permissions.ComponentA.read || permissions.ComponentA.write || permissions.ComponentB.read || ...">
  <!-- UI stuff showing outputs from SomeServiceA.method1() and SomeServiceB.method2() >
</div>
```

The above may often lead to complicate condition checking, and developers having to know before-hand and code scattered authorization statements in the Views. As UI is basically consuming and invoking `Service` methods, we may just leave `Service` methods tell us whether our current permissions are enough to go on with them.

So the alterntive approach would be:
```html
<div *ngIf="SomeServiceA.method1.allowed && SomeServiceB.method2.allowed">
  <!-- UI stuff showing outputs from SomeServiceA.method1() and SomeServiceB.method2() >
</div>
```

```ts
// some-service-a.service.ts
@Scope(Scope.SOME_BACKEND_SCOPE)
@Injectable({                                                                                        
  providedIn: WhateverModule                                                                              
})                                                                                                   
export class SomeServiceA {                                                                      
  
  @requires_permissions(Permissions.READ)
  method1(...) {

  }
``` 

For back-driven services this could also be offloaded to the back-end rest api endpoints, which already know their scope and permissions. By implementing `HTTP OPTIONS` call:
```http
# A user with RGW read only permissions issues an `OPTIONS /api/rgw/buckets` 
HTTP/1.1 204 No Content
Allow: OPTIONS, GET, HEAD
...
X-Dashboard-Scope: RGW
```

And, in the front-end:
```ts
// smart-api.service.ts
export abstract class SmartAPIService {
  private url:string;
                                                                      
  constructor(private http: HttpClient) {
    response = this.options(); // if Forbidden, all CRUD methods are disabled
    this.scope = response.headers.get('X-Dashboard-Scope');
    this.permissions = response.headers.get('Allows');
  }
}

// api/rgw-bucket.service.ts
@Injectable({                                                                                        
  providedIn: ApiModule                                                                              
})                                                                                                   
export class RgwBucketService extends SmartAPIService {                                                                      
  private url = 'api/rgw/bucket';
  // Inheritance would implement scope & permission from child url member
  }
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html:
```diff
@@ -92,13 +92,27 @@
       </li>
+      <li routerLinkActive="active"
+          class="tc_submenuitem tc_submenuitem_cluster_inventory"
+          *ngIf="permissions.hosts.read">
```
Not sure if we should hide it if the orchestrator is not available.
For example for the Prometheus configuration we hide the menus.

bk201 replied:
These two menu items are hidden when the user has no `HOSTS.read` permission. If orchestrator is unavailable, the information about how to enable it is displayed.

bk201 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -0,0 +1,82 @@
+  private setComplexValidators() {
+    this.hostForm.get('hostname').setValidators([
+      this.hostForm.get('hostname').validator,
+      CdValidators.custom('uniqueName', (hostname: string) => {
```
Thanks, fixed in 9d9a79c

bk201 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/inventory/inventory.component.html:
```diff
@@ -0,0 +1,22 @@
+  <a href="{{ docsUrl }}"
+     target="_blank">documentation</a> on how to
+  configure and enable the orchestrator functionality.
+</cd-info-panel>
```
Fixed in 9d9a79c

bk201 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/services.component.html:
```diff
@@ -0,0 +1,17 @@
+  <a href="{{ docsUrl }}"
+     target="_blank">documentation</a> on how to
+  configure and enable the orchestrator functionality.
+</cd-info-panel>
```
Fixed in 9d9a79c

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -1,12 +1,104 @@
+
+
+def merge_hosts_by_hostname(ceph_hosts, orch_hosts):
+    """Merge Ceph hosts with orchestrator hosts by hostnames.
```
@jmolmo + @tserong In the past, I've seen uses thatused different host aliases for different network interfaces of physical hosts. At least for DeepSea, I'm expecting this to break in some corner cases. Are there similar uses cases with ceph-ansible?

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/orchestrator.py:
```diff
@@ -0,0 +1,53 @@
+        return OrchClient.instance().status()
+
+
+@ApiController('/orchestrator/inventory', Scope.HOSTS)
```
hm, what do you think of merging the orchestrator inventory with `ceph device ls`? similar to hosts?

bk201 replied:
@votdev raise this question before. There is an issue for tracking `device ls` feature: https://tracker.ceph.com/issues/39352 and it is still pending for discussion.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/orchestrator.py:
```diff
@@ -0,0 +1,53 @@
+    @raise_if_no_orchestrator
+    def list(self, hostname=None):
+        orch = OrchClient.instance()
+        return [service.to_json() for service in orch.services.list(None, None, hostname)]
```
Does this overlap with the OSD list? I wonder if it would make sense to somehow unify the view of the orchestrator's services with the rest of the dashboard?

bk201 replied:
Yeah, Orchestrator service entities should be hidden behind various components of Dashboard. We already have an OSD list and if we want to manipulate OSD, we then just leverage orchestrator API calls in the backend. This also applies to MON, MDS, ..etc.

We don't have an issue on the tracker to display all orchestrator services (but we have one pending item on the [etherpad](https://pad.ceph.com/p/ceph-dashboard-octopus-priorities) during f2f, item 22.6). I implemented it just because orchestrator CLI has it.

Do users have the need to see *all* services or services of a host? Also, orchestrator services display slightly more information. For example, when using a Rook backend it might be useful to know the container id. Of course, this can also be done by showing additional info on MON/MDS/MON components too.

tspmelo commented:
jenkins test dashboard

bk201 commented:
Add DNM label. Teuthology tests in this PR depends on https://github.com/ceph/ceph/pull/29595 t o be merged.

bk201 commented:
jenkins test dashboard please

bk201 commented:
> Please fix the commits.
> Some messages don't have the component information.
> Some commits are reverting changes made in previous commits, please squash those.

Thanks, rebased and squashed.

bk201 commented:
jenkins test dashboard

bk201 commented:
jenkins test dashboard

sebastian-philipp commented:
jenkins test dashboard

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-09-07_14:44:52-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-08_13:41:17-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-smithi/
